lua/plenary/path.lua

```diff
@@ -198,6 +198,57 @@ function Path:expand()
   return expanded and expanded or error("Path not valid")
 end
 
+function Path:make_relative()
+  if self.filename:sub(1, #self._cwd) == self._cwd  then
+    local offset =  0
+    -- if  self._cwd does ends in the os separator, we need to take it off
+    if self._cwd:sub(#self._cwd, #self._cwd) ~= path.separator then
+      offset = 1
+    end
+
+    self.filename = self.filename:sub(#self._cwd + 1 + offset, #self.filename)
+  end
+
+  return self.filename
+end
+
+function Path:normalize()
+  self:make_relative()
+  -- Substitute home directory w/ "~"
+  self.filename = self.filename:gsub("^" .. path.home, '~', 1)
+  -- Remove double path seps, it's annoying
+  self.filename = self.filename:gsub(path.sep .. path.sep, path.sep)
+  print(self.filename)
+
+  return self.filename
+end
+
+local shorten = (function()
+  if jit then return function(filename)
+    local ffi = require('ffi')
+    ffi.cdef [[
+    typedef unsigned char char_u;
+    char_u *shorten_dir(char_u *str);
+    ]]
+
+    if not filename then
+      return filename
+    end
+
+    local c_str = ffi.new("char[?]", #filename + 1)
+    ffi.copy(c_str, filename)
+    return ffi.string(ffi.C.shorten_dir(c_str))
+    end
```

Comment:
Thats not 100% correct. You only wanna `return function(filename)` everything after cdef. You don't need to do the ffi stuff every time.

Reply:
Ah ok. I don't really know what ffi is so this is all very helpful. Thanks :)

Reply:
https://luajit.org/ext_ffi.html

Basically `ffi` allows us to call c functions. We have to define them like they are defined in a header file and then we can call them. We might have to load a shared lib to make this work. E.g. If we don't use some system functions or vim function like here (`shorten_dir` is part of vim). 

E.g. if you use `telescope-fzy-native` its basically a c lib which implements everything for us and we just call the c functions to achieve super sonic speed.